File: hello-nodejs/src/server/mmapCache.js
// File-backed cache for a single audio stream.
import * as fs from 'fs';
import * as path from 'path';

const MAX_CACHE_SIZE = 8 * 1024 * 1024 * 1024; // 8 GB

export class MmapCache {
  #filePath;
  #fd = null;
  #fileSize = 0;
  #active = false;

  constructor(filePath) {
    this.#filePath = filePath;
  }

  create() {
    const dir = path.dirname(this.#filePath);
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
    if (fs.existsSync(this.#filePath)) fs.unlinkSync(this.#filePath);
    this.#fd = fs.openSync(this.#filePath, 'w+');
    this.#fileSize = 0;
    this.#active = true;
  }

  /** @param {number} offset @param {Buffer} data @returns {number} */
  write(offset, data) {
    if (!this.#active || this.#fd === null) this.create();
    if (offset + data.length > MAX_CACHE_SIZE) throw new Error('Exceeds max cache size');
    fs.writeSync(this.#fd, data, 0, data.length, offset);
    if (offset + data.length > this.#fileSize) this.#fileSize = offset + data.length;
    return data.length;
  }

  /** @param {number} offset @param {number} length @returns {Buffer} */
  read(offset, length) {
    if (!this.#active || this.#fd === null) {
      if (!fs.existsSync(this.#filePath)) return Buffer.alloc(0);
      this.#fd = fs.openSync(this.#filePath, 'r');
      this.#fileSize = fs.fstatSync(this.#fd).size;
      this.#active = true;
    }
    if (offset >= this.#fileSize) return Buffer.alloc(0);
    const actual = Math.min(length, this.#fileSize - offset);
    const buf = Buffer.alloc(actual);
    fs.readSync(this.#fd, buf, 0, actual, offset);
    return buf;
  }

  /** @param {number} finalSize */
  finalize(finalSize) {
    if (this.#fd !== null) fs.ftruncateSync(this.#fd, finalSize);
    this.#fileSize = finalSize;
  }

  close() {
    if (this.#fd !== null) {
      fs.closeSync(this.#fd);
      this.#fd = null;
    }
    this.#active = false;
  }
}

File: hello-dart/lib/server/mmap_cache.dart
import 'dart:io';

class MmapCache {
  final String path;
  RandomAccessFile? _file;
  int _size = 0;
  bool _open = false;

  MmapCache(this.path);

  bool create() {
    close();
    final f = File(path);
    final dir = f.parent;
    if (!dir.existsSync()) {
      dir.createSync(recursive: true);
    }
    if (f.existsSync()) {
      f.deleteSync();
    }
    _file = f.openSync(mode: FileMode.write);
    _size = 0;
    _open = true;
    return true;
  }

  int write(int offset, List<int> data) {
    if (!_open || _file == null || data.isEmpty) return 0;

    final needed = offset + data.length;
    if (needed > _size) {
      _file!.truncateSync(needed);
      _size = needed;
    }

    _file!.setPositionSync(offset);
    _file!.writeFromSync(data);
    return data.length;
  }

  List<int> read(int offset, int length) {
    if (!_open || _file == null || offset >= _size || length <= 0) return [];

    final actual = length < (_size - offset) ? length : (_size - offset);
    _file!.setPositionSync(offset);
    return _file!.readSync(actual);
  }

  bool finalize(int finalSize) {
    if (!_open || _file == null) return false;
    _file!.truncateSync(finalSize);
    _file!.flushSync();
    _size = finalSize;
    return true;
  }

  void close() {
    if (_file != null) {
      _file!.closeSync();
      _file = null;
    }
    _open = false;
  }

  void delete() {
    close();
    final f = File(path);
    if (f.existsSync()) {
      f.deleteSync();
    }
  }

  int get size => _size;
  bool get isOpen => _open;
}
